daemon/state.go

```diff
@@ -47,67 +48,73 @@ func (d *Daemon) SyncState(dir string, clean bool) error {
 	possibleEPs := readEPsFromDirNames(dir, eptsID)
 
 	if len(possibleEPs) == 0 {
-		log.Debug("No old endpoints found.")
+		log.Info("No old endpoints found.")
 		return nil
 	}
 
 	for _, ep := range possibleEPs {
 		log.Debugf("Restoring endpoint ID %d", ep.ID)
 
 		if err := d.syncLabels(ep); err != nil {
-			log.Warningf("Unable to restore endpoint %+v: %s", ep, err)
+			log.Warningf("Unable to restore endpoint %d: %s", ep.ID, err)
 			continue
 		}
 
+		ep.Mutex.Lock()
 		if d.conf.KeepConfig {
 			ep.SetDefaultOpts(nil)
 		} else {
 			ep.SetDefaultOpts(d.conf.Opts)
 		}
 
-		if buildSuccess := <-ep.Regenerate(d); !buildSuccess {
-			continue
-		}
-
 		endpointmanager.Insert(ep)
 		restored++
 
-		log.Infof("Restored endpoint: %d", ep.ID)
+		ep.Mutex.Unlock()
 	}
 
 	log.Infof("Restored %d endpoints", restored)
 
 	if clean {
-		d.cleanUpDockerDandlingEndpoints()
+		d.cleanUpDockerDanglingEndpoints()
 	}
 
 	endpointmanager.Mutex.Lock()
+	var wg sync.WaitGroup
+	wg.Add(len(endpointmanager.Endpoints))
 	for k := range endpointmanager.Endpoints {
 		ep := endpointmanager.Endpoints[k]
-		if err := d.allocateIPs(ep); err != nil {
-			log.Errorf("Failed while reallocating ep %d's IP addresses: %s. Endpoint won't be restored", ep.ID, err)
-			d.deleteEndpoint(ep)
-			continue
-		}
+		go func(ep *endpoint.Endpoint, wg *sync.WaitGroup) {
+			if err := d.allocateIPs(ep); err != nil {
+				log.Errorf("Failed while reallocating ep %d's IP addresses: %s. Endpoint won't be restored", ep.ID, err)
+				d.deleteEndpoint(ep)
+				wg.Done()
+				return
+			}
 
-		log.Infof("EP %d's IP addresses successfully reallocated", ep.ID)
-		if buildSuccess := <-ep.Regenerate(d); !buildSuccess {
-			log.Warningf("Failed while regenerating endpoint %d: %s", ep.ID, err)
-			continue
-		}
-		ep.Mutex.RLock()
-		epID := ep.ID
-		if ep.SecLabel != nil {
-			epLabels := ep.SecLabel.DeepCopy()
-			ep.Mutex.RUnlock()
-			d.events <- *events.NewEvent(events.IdentityAdd, epLabels)
-		} else {
-			ep.Mutex.RUnlock()
-		}
-		log.Infof("Restored endpoint %d", epID)
+			log.Infof("EP %d's IP addresses successfully reallocated", ep.ID)
+			if buildSuccess := <-ep.Regenerate(d); !buildSuccess {
+				log.Warningf("Failed while regenerating endpoint %d: %s", ep.ID, err)
+				wg.Done()
+				return
+			}
+			ep.Mutex.RLock()
+			epID := ep.ID
+			if ep.SecLabel != nil {
+				epLabels := ep.SecLabel.DeepCopy()
+				ep.Mutex.RUnlock()
+				d.events <- *events.NewEvent(events.IdentityAdd, epLabels)
+			} else {
+				ep.Mutex.RUnlock()
+			}
+			log.Infof("Restored endpoint %d", epID)
```

Comment:
This is the most important info message, we may want to report the IP addresses as well.